<a href="https://colab.research.google.com/github/Crisosth/IA/blob/main/AtividadeRNA_GustavoCrisosth.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
#Importação da biblioteca pandas
import pandas as pd

In [ ]:
#base de dados =  previsores + classe

#criação atributos previsores - 7 colunas x 900 registros
previsores = pd.read_csv('passas_entradas.csv')

#criação classe - 1 coluna x 900 registros
classe = pd.read_csv('passas_saida.csv')

In [ ]:
#importação da função que faz a divisão da base de dados
from sklearn.model_selection import train_test_split

In [ ]:
#divisão da base de dados - treinamento(70% = 630 registros) - teste (30% = 270 registros)
#criação das variaveis que vão receber as partes da base de dados divididas
#previsores_treinamento = 7 colunas x 630
#previsores_teste = 7 colunas x 270 registros
#classe_treinamento = 1 coluna x 630 registros
#classe_teste = 1 coluna x 270 registros

previsores_treinamento, previsores_teste, classe_treinamento, classe_teste = train_test_split(previsores, classe, test_size=0.30)

In [ ]:
previsores_treinamento

Area  MajorAxisLength  MinorAxisLength  Eccentricity  ConvexArea  \
604   94211       450.004617       269.286569      0.801191       96340   
322   67798       418.408134       207.825511      0.867919       69356   
498   82886       424.822709       253.171548      0.803024       85879   
753  124630       585.093803       275.493613      0.882211      128308   
454   79055       395.368667       258.275952      0.757139       82562   
..      ...              ...              ...           ...         ...   
455   79057       436.390047       236.895393      0.839828       82642   
203   57785       314.430977       236.183035      0.660138       58867   
631   97583       522.786556       241.940773      0.886468      101231   
843  166275       602.307775       356.236287      0.806340      168292   
790  137583       649.541485       273.260282      0.907201      142650   

       Extent  Perimeter  
604  0.716848   1194.631  
322  0.691027   1073.138  
498  0.648281   1163.528  
753  0.681832   1485.990  
454  0.719506   1114.488  
..        ...        ...  
455  0.654597   1148.146  
203  0.732866    903.769  
631  0.725983   1298.731  
843  0.719282   1580.961  
790  0.731638   1590.354  

[630 rows x 7 columns]

In [ ]:
# importação da biblioteca keras
import keras

# importação do modelo sequencial
# sequencia de camadas: Entrada-Oculta-Saída
from keras.models import Sequential

# importação da classe de camadas densas
from keras.layers import Dense

In [ ]:
# criando a rede neural do tipo sequencial(objeto)
# nome = classificador
classificador = Sequential()

# adicionando a primeira camada oculta
classificador.add(Dense(units= 32, activation= 'relu',
                        kernel_initializer = 'random_uniform', input_dim= 7))
# adicionando a segunda camada oculta
classificador.add(Dense(units = 32, activation= 'relu', kernel_initializer= 'random_uniform'))

# adicionando a terceira camada oculta
classificador.add(Dense(units = 16, activation= 'relu', kernel_initializer= 'random_uniform'))

In [ ]:
# adicionando a camada de saída
classificador.add(Dense(units=1, activation= 'sigmoid'))

In [ ]:
#compilando a rede neural com o metodo compile
otimizador = keras.optimizers.Adam(learning_rate= 0.001, weight_decay= 0.0021, clipvalue=5.5)
classificador.compile(optimizer = otimizador, loss= 'binary_crossentropy', metrics = ['binary_accuracy'])
#classificador.compile(optimizer = 'adam', loss = 'binary_crossentropy', metrics = ['binary_accuracy'])


In [ ]:
# treinando a Rede Neural
classificador.fit(previsores_treinamento, classe_treinamento, batch_size= 8, epochs = 190)

Epoch 1/190
79/79 [==============================] - 0s 4ms/step - loss: 0.4369 - binary_accuracy: 0.8111
Epoch 2/190
79/79 [==============================] - 0s 3ms/step - loss: 0.4268 - binary_accuracy: 0.8365
Epoch 3/190
79/79 [==============================] - 0s 3ms/step - loss: 0.4881 - binary_accuracy: 0.8048
Epoch 4/190
79/79 [==============================] - 0s 3ms/step - loss: 0.4470 - binary_accuracy: 0.8127
Epoch 5/190
79/79 [==============================] - 0s 4ms/step - loss: 0.4267 - binary_accuracy: 0.8127
Epoch 6/190
79/79 [==============================] - 0s 3ms/step - loss: 0.4233 - binary_accuracy: 0.8302
Epoch 7/190
79/79 [==============================] - 0s 3ms/step - loss: 0.4184 - binary_accuracy: 0.8317
Epoch 8/190
79/79 [==============================] - 0s 3ms/step - loss: 0.4458 - binary_accuracy: 0.8111
Epoch 9/190
79/79 [==============================] - 0s 3ms/step - loss: 0.4287 - binary_accuracy: 0.8222
Epoch 10/190
79/79 [==========================

In [ ]:
# executando a Rede Neural com a Base de Dados
previsoes = classificador.predict(previsores_teste)

9/9 [==============================] - 0s 3ms/step


In [ ]:
previsoes

array([[0.7698403 ],
       [0.9549105 ],
       [0.3292462 ],
       [0.54809374],
       [0.05574378],
       [0.03490391],
       [0.76904744],
       [0.31372964],
       [0.1500929 ],
       [0.7639432 ],
       [0.28489435],
       [0.28702283],
       [0.15527995],
       [0.77528036],
       [0.29623255],
       [0.08367134],
       [0.17141461],
       [0.13675371],
       [0.30066034],
       [0.20163245],
       [0.9246989 ],
       [0.97123474],
       [0.7398855 ],
       [0.06032872],
       [0.13481408],
       [0.24691345],
       [0.05907571],
       [0.9827342 ],
       [0.92382395],
       [0.21274167],
       [0.20553549],
       [0.1764898 ],
       [0.04768891],
       [0.99151844],
       [0.99851406],
       [0.10414848],
       [0.5927533 ],
       [0.10901113],
       [0.33438927],
       [0.8640167 ],
       [0.98242265],
       [0.05153272],
       [0.35773227],
       [0.9454125 ],
       [0.1857343 ],
       [0.08167434],
       [0.9326079 ],
       [0.946

In [ ]:
#convertendo as previsoes da Rede Neural para V ou F
previsoes = (previsoes > 0.5)
previsoes

array([[ True],
       [ True],
       [False],
       [ True],
       [False],
       [False],
       [ True],
       [False],
       [False],
       [ True],
       [False],
       [False],
       [False],
       [ True],
       [False],
       [False],
       [False],
       [False],
       [False],
       [False],
       [ True],
       [ True],
       [ True],
       [False],
       [False],
       [False],
       [False],
       [ True],
       [ True],
       [False],
       [False],
       [False],
       [False],
       [ True],
       [ True],
       [False],
       [ True],
       [False],
       [False],
       [ True],
       [ True],
       [False],
       [False],
       [ True],
       [False],
       [False],
       [ True],
       [ True],
       [ True],
       [ True],
       [False],
       [False],
       [False],
       [ True],
       [ True],
       [False],
       [ True],
       [False],
       [ True],
       [ True],
       [False],
       [ True],
       [

In [ ]:
# avaliando o acerto - comparar previsoes e classe_teste
#importando o método de comparação
from sklearn.metrics import accuracy_score

#avaliando a precisão das previsoes com a classe teste
precisao = accuracy_score(classe_teste, previsoes)

In [ ]:
#mostrando a precisao
precisao

0.9037037037037037

In [ ]:
# Avaliandoi o acerto - comparar previsores e classe_teste
#usando recursos do Keras
resultado= classificador.evaluate(previsores_teste, classe_teste)
resultado

9/9 [==============================] - 0s 2ms/step - loss: 0.2816 - binary_accuracy: 0.9037


[0.2815520167350769, 0.9037036895751953]

In [ ]:
loss, accuracy = classificador.evaluate(previsores_teste, classe_teste)
print('Loss: %.2f' % loss)
print('Accuracy: %.2f' % (accuracy*100))

9/9 [==============================] - 0s 4ms/step - loss: 0.2816 - binary_accuracy: 0.9037
Loss: 0.28
Accuracy: 90.37


# Percentual de acerto final de 90.37%

##Parametros utilizados:
####•3 camadas ocultas com ativação reLu com 32, 32 e 16 neuronios e 1 camada de saída com ativação sigmoid
####•Otimizador adam com weight_decay = 0.0021 e clipvalue = 5.5
####•Batch size de 8 e 190 epocas
